File: bindings/rust/src/safe_calculator.rs
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>

//! Safe expression evaluator via libproven FFI.
//!
//! Evaluates mathematical expressions with overflow and division-by-zero
//! protection. All operations delegate to Idris 2 verified code.

use crate::core::{self, Result};
use crate::ffi;

/// Safe calculator for mathematical expressions.
pub struct SafeCalculator;

impl SafeCalculator {
    /// Evaluate a mathematical expression string.
    ///
    /// Supports basic arithmetic operators (+, -, *, /) with proper
    /// precedence. Returns `Err` for division by zero, overflow, or
    /// invalid syntax.
    ///
    /// # Examples
    ///
    /// ```ignore
    /// let result = SafeCalculator::eval("2 + 3 * 4");
    /// // result == Ok(14.0)
    /// ```
    pub fn eval(expression: &str) -> Result<f64> {
        let bytes = expression.as_bytes();
        // SAFETY: We pass a valid pointer and length from the string's bytes.
        // The FFI function only reads from the pointer within the given length.
        let result = unsafe {
            ffi::proven_calculator_eval(bytes.as_ptr(), bytes.len())
        };
        core::float_result_to_result(result)
    }
}

File: bindings/zig/src/safe_calculator.zig
// SPDX-License-Identifier: PMPL-1.0-or-later
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
//
// Proven SafeCalculator - FFI bindings to libproven expression evaluation.
// All computation is performed in verified Idris 2 code via libproven.

const std = @import("std");
const c = @cImport(@cInclude("proven.h"));

/// Error types for calculator operations.
pub const CalcError = error{
    ParseFailure,
    DivisionByZero,
    ProvenError,
};

/// Evaluate an arithmetic expression safely via libproven.
/// Supports +, -, *, /, parentheses, negative and decimal numbers.
pub fn eval(expression: []const u8) CalcError!f64 {
    const result = c.proven_calculator_eval(expression.ptr, expression.len);
    return switch (result.status) {
        c.PROVEN_OK => result.value,
        c.PROVEN_ERR_PARSE_FAILURE => error.ParseFailure,
        c.PROVEN_ERR_DIVISION_BY_ZERO => error.DivisionByZero,
        else => error.ProvenError,
    };
}

test "eval" {
    try std.testing.expectApproxEqAbs(@as(f64, 7.0), try eval("3 + 4"), 0.001);
    try std.testing.expectApproxEqAbs(@as(f64, 14.0), try eval("2 * (3 + 4)"), 0.001);
    try std.testing.expectError(error.DivisionByZero, eval("1 / 0"));
    try std.testing.expectError(error.ParseFailure, eval("not math"));
}
